src/providers/http.rs

```diff
@@ -0,0 +1,169 @@
+        }
+    }
+}
+
+#[derive(Deserialize, Serialize, Debug, Clone)]
+#[serde(deny_unknown_fields, default)]
+pub struct HttpConfig {
+    url: Option<Url>,
+    request: RequestConfig,
+    poll_interval_secs: u64,
+    #[serde(flatten)]
+    tls_options: Option<TlsOptions>,
+}
+
+impl Default for HttpConfig {
+    fn default() -> Self {
+        Self {
+            url: None,
+            request: RequestConfig::default(),
+            poll_interval_secs: 30,
+            tls_options: None,
+        }
+    }
+}
+
+async fn http_request(
+    url: &Url,
+    tls_options: &Option<TlsOptions>,
+    headers: &IndexMap<String, String>,
+) -> Result {
+    let tls_settings = TlsSettings::from_options(tls_options).map_err(|_| "Invalid TLS options")?;
+    let http_client = HttpClient::<Body>::new(tls_settings).map_err(|_| "Invalid TLS settings")?;
+
+    // Build HTTP request.
+    let mut builder = http::request::Builder::new().uri(url.to_string());
+
+    // Augment with headers. These may be required e.g. for authentication to
+    // private endpoints.
+    for (header, value) in headers.iter() {
+        builder = builder.header(header.as_str(), value.as_str());
+    }
+
+    let request = builder
+        .body(Body::empty())
+        .map_err(|_| "Couldn't create HTTP request")?;
+
+    info!(
+        message = "Attempting to retrieve configuration.",
+        url = ?url.as_str()
+    );
+
+    let response = http_client.send(request).await.map_err(|err| {
+        let message = "HTTP error";
+        error!(
+            message = ?message,
+            error = ?err,
+            url = ?url.as_str());
+        message
+    })?;
+
+    info!(message = "Response received.", url = ?url.as_str());
+
+    let body = hyper::body::to_bytes(response.into_body())
+        .await
+        .map_err(|err| {
+            let message = "Error interpreting response.";
+            let cause = err.into_cause();
+            error!(
+                    message = ?message,
+                    error = ?cause);
+
+            message
+        })?;
+
+    Ok(String::from_utf8_lossy(body.as_ref()).to_string())
+}
+
+#[async_trait::async_trait]
+#[typetag::serde(name = "http")]
+impl ProviderConfig for HttpConfig {
+    async fn build(&mut self, signal_handler: &mut signal::SignalHandler) -> Result {
+        let url = self
+            .url
+            .take()
+            .ok_or("URL is required for the `http` provider.")?;
+
+        let tls_options = self.tls_options.take();
+        let poll_interval_secs = self.poll_interval_secs;
+        let request = self.request.clone();
+
+        let (signal_tx, signal_rx) = mpsc::channel(2);
+
+        let mut shutdown_rx = signal_handler
+            .with_shutdown(ReceiverStream::new(signal_rx))
+            .subscribe();
+
+        // Attempt to retrieve the initial configuration, then poll for changes.
+        http_request(&url, &tls_options, &request.headers).await.map(|config| {
```

Comment:
This seems like an overly confusing way to structure this. Why not something like:

```rust
let config = http_request(...).await?;
tokio::spawn(...);
Ok(config)
```

Reply:
Yep, much clearer. Flattened in 9cc1b33cb3bb3b2b592a7ff9fb51c5df6706536c.